### ISSUE TITLE
Taxes not including promotions with whole-order adjustments

### ISSUE THREAD

**Issue by luisenrike** [spree/spree#9231]
## Context
In an order with promotions that have whole-order adjustments, taxes are not including the discounts.

## Expected Behavior
Taxes should be calculated after whole-order promotions have been applied to the order.

## Actual Behavior
Taxes are not including promotions discounts.

## Steps to Reproduce

1. Create a promotion with a whole-order adjustment and a flat percent discount
2. Add an item to the cart and add that promotion
3. Go to checkout and add an address that it's part of a tax zone.

![spree_administration__promotions](https://user-images.githubusercontent.com/12720974/52229045-e5a1f700-2881-11e9-96fe-11b5dca5f3b5.png)

![shopping_cart_-_spree_demo_site](https://user-images.githubusercontent.com/12720974/52229251-7aa4f000-2882-11e9-97e0-e6895e3fe544.png)

![spree_demo_site](https://user-images.githubusercontent.com/12720974/52229406-d40d1f00-2882-11e9-947d-d9ec73aa070e.png)

## Your Environment

* Version used:
* Gemfile and Gemfile.lock as text in a Gist:
* Any relevant stack traces ("Full trace" preferred):

Default Spree 3.5 and Rails 5.1.6 installation.

/cc @damianlegawiec @mamhoff

**Comment by MatthewKennedy** [comment]
This is a tricky issue, as what if you sell items that are zero rated and have a mixed cart?

Item 1 €25 - includes 20% VAT (€20.83 with €4.17 VAT)
Item 2 €25 - includes 20% VAT (€20.83 with €4.17 VAT)
Item 3 €25 - includes 20% VAT (€20.83 with €4.17 VAT)
Item 4 €25 - includes 0% VAT   (€25 with €0 VAT)

Line Item total €87.49
VAT €12.51 
Subtotal €100

Whole Order Discount 50%
Total €50 <- of that how much is tax?
€43.75 with €6.25 VAT

But if you just took the total then applied a whole order discount of 50% then did the tax and run that through a VAT calculator €41.67 with €8.33 VAT, so the whole order discount needs to take into account at the tax that is in the total, they may be zero rated items.

I'll tell you what I do, though it might not be entirely legal in some countries, depending on the rules.
![screenshot 2019-02-04 at 21 21 43](https://user-images.githubusercontent.com/1240766/52238156-46055880-28c3-11e9-95be-efed456cd3d9.jpg)
And then the customer is emailed the correct invoice from a dedicated accounting software.

**Comment by MatthewKennedy** [comment]
Is that a whole order discount or line item discount?

**Comment by MatthewKennedy** [comment]
Yeah, but the Sales Tax should be $1.60 not $3.20

Your final sale price on that $15.99 USD product is $7.99 after the discount, but your giving the government the $3.20 tax as if you sold it at $15.99

**Comment by luisenrike** [comment]
Yes, you're right, is the same problem. Taxes should be calculated after discounts.

**Comment by virajrch** [comment]
For those who need a solution before this issue is fixed, I have found a trick that can achieve this.
The solution is to create a "line item adjustment" instead with a rule of type "Product(s)" and then keeping the products field blank and choose "none" for "Order must contain x amount of these products". This automatically recalculates the tax when the promo code is applied.

Screenshot:
![image](https://user-images.githubusercontent.com/2718778/57017838-8ebc7b80-6c3e-11e9-95b1-dcd184822b98.png)

**Comment by MatthewKennedy** [comment]
@virajrch That solution works and is the right way to go, but, unfortunately there is a small issue with "line item adjustment" where it is not checked for applicability past the cart stage. as mentioned #9375 This issue is addressed in. PR #9380 

**Whole Order Discount Tax Issue**
When thinking about whole order discounts and the issues raised with tax it would simply be best to use line-item discounts in all scenarios if you charge sales taxes at all.

There is no easy solution to this whole order discount and the tax issues, because even if you worked out a unique tax percentage on each cart and its mix of taxed and none taxed items, and then worked out the amount within the discount that is tax based on the unique tax percentage from taxed and none taxed items in a cart, it would still generate an incorrect invoice showing tax across all items at a lower rate, its just not a good idea.

Personally I would not invest developer time in trying to make whole order discounts act on a per line item basis, and simply use the per line item discounts as you suggested, it makes sense. 

Maybe put a warning in Promotions section and the documentation about using whole order discounts if you charge sales taxes.

**Comment by Quintasan** [comment]
@MatthewKennedy how would you approach a flat rate whole-order discount then? If, for example, I have a cart worth 50 EUR and apply a promo code that gives a flat 50 EUR discount on the WHOLE order, Spree still calculates the VAT from the price before the discount. Is there any way I can apply the promo code before Spree calculates VAT?

**Comment by MatthewKennedy** [comment]
@Quintasan if you apply the code in the cart is the VAT still showing when you go into the checkout?

**Comment by Quintasan** [comment]
Yup, it even gets saved in the database when the order is placed.

wt., 13 sie 2019, 15:07 użytkownik Matthew Kennedy <notifications@github.com>
napisał:

**Comment by MatthewKennedy** [comment]
@Quintasan The €50 cart, does this have multiple products each with tax included?

**Comment by MatthewKennedy** [comment]
@Quintasan You will need to use line item based discounts if you want to accurately reflect tax after discounts.

**Comment by Quintasan** [comment]
It has one product worth 50 EUR, after reduction the order would be worth 0
EUR and VAT for that should be 0 as well.

wt., 13 sie 2019, 15:30 użytkownik Matthew Kennedy <notifications@github.com>
napisał:

**Comment by Quintasan** [comment]
@MatthewKennedy I was doing testing with one product in cart, every product in my shop has tax included.

Is it even possible to get the behaviour I want in Spree?

**Comment by MatthewKennedy** [comment]
@Quintasan I'd say no it's not possible.

**Comment by MatthewKennedy** [comment]
@Quintasan In other platforms you have come across, does a fixed amount "whole order discount" apply to just cart items, not shipping?

**Comment by Quintasan** [comment]
@MatthewKennedy I wouldn't say it applies to shipping but my shop configuration has no shipping fees. I still think that behaviour like this should be treated as a bug:

Before checkout
![obraz](https://user-images.githubusercontent.com/6539261/63003243-0f168d80-be78-11e9-81b3-4fe612fb88ab.png)

After checkout
![obraz](https://user-images.githubusercontent.com/6539261/63003364-438a4980-be78-11e9-943c-6da3b6ffc1d4.png)

In admin
![obraz](https://user-images.githubusercontent.com/6539261/63003435-674d8f80-be78-11e9-8a6e-d40b0bd45471.png)

 I mean, Spree calculates VAT BEFORE the discount gets applied and I can't exactly use per line item based discount if I want to give my users a flat discount from the total order price

**Comment by MatthewKennedy** [comment]
@Quintasan To properly use a fixed amount whole order discount with tax you need to have the following take place:

1. Find what percentage of the order total each line item has, not including shipping or tax.
2. Take that percentage and calculate what amount of the discounts monetary value each line item should be allocated.
3. Remove the allocate discount amount from each line item.
4. Update order with the new line items values.
5. Add tax on the line items new vales where necessary.

**Comment by sacha75** [comment]
I opened a PR with a simple resolution for this matter, I hope this could be merged as we are currently using custom methods to get correct result on our spree.
https://github.com/spree/spree/pull/10116

**Comment by dylanfisher** [comment]
I realize this is an old issue, but thought I'd share what I found to be the easiest way to handle this issue. This involves defining a custom tax calculator and offsetting the tax rate of each line item by the number of items receiving the promotion. My calculator looks something like this:

```ruby
module Spree
  class Calculator::CustomTaxCalculator < Calculator::DefaultTax
    def self.description
      "Custom Tax Calculator"
    end

    def compute_line_item(line_item)
      return super if rate.included_in_price
      return super unless line_item.needs_special_tax_calculation? # `needs_special_tax_calculation?` is going to be based on whatever special promotion rules apply to your line items

      promotions = line_item.order.adjustments.select { |a| ['My Special Discount'].include?(a.source.promotion.name) }
      promotion_total = promotions.collect(&:amount).sum
      taxable_promotion_difference = promotion_total * rate.amount
      taxable_promotion_difference_per_item = taxable_promotion_difference / line_item.order.line_items.size

      round_to_two_places(line_item.discounted_amount * rate.amount + taxable_promotion_difference_per_item)
    end
  end
end
```
